### PR TITLE
ByteScheduler support for PyTorch

### PR DESCRIPTION
**Pull request by yhpeng-git**
This is a PR for adding ByteScheduler support to BytePS in PyTorch. ByteScheduler enables overlapping gradient push-pull with both backward computation and forward computation by removing the global barrier between steps. It maintains correct dependencies, e.g., the forward computation of a neural layer will not start until the parameter of this layer is updated. Hence it can further improves training performance beyond BytePS.

To use it, just wrap BytePS optimizer using ScheduledOptimizer with a few other arguments. So far ByteScheduler for PyTorch supports SGD, Adam and RMSprop optimizers and all of them are tested. A full example is provided in file "benchmark_bytescheduler.py". Set BYTESCHEDULER_DEBUG=1 to see debugging output to verify if it is enabled.

### CODE REVIEW

**Comment by bobzhuyb** [comment]
@yhpeng-git Can you change it a bit? Let users just import byteps.bytescheduler.torch.optimizer as bps, then directly use bps to wrap PyTorch's original optimizer. No need to wrap bps and then wrap bsc again.

**Comment by yhpeng-git** [comment]
> @yhpeng-git Can you change it a bit? Let users just import byteps.bytescheduler.torch.optimizer as bps, then directly use bps to wrap PyTorch's original optimizer. No need to wrap bps and then wrap bsc again.

**Comment by yhpeng-git** [comment]
I changed the usage way a bit. Now users only need to  import byteps.bytescheduler.torch.optimizer as bps and add two additional arguments when wrapping the PyTorch optimizer.

**Comment by ymjiang** [comment]
Hi @yhpeng-git , could you please add a README under byteps/bytescheduler? It would be nice if the README contains the following information: the rationale of bytescheduler, the basic usage (how to use in user codes), a tutorial (like [this](https://github.com/bytedance/byteps/blob/master/docs/step-by-step-tutorial.md)).

**Comment by ymjiang** [comment]
@yhpeng-git Please also rebase to the latest master branch.

**Comment by yhpeng-git** [comment]
@ymjiang Added the README and rebased to lastest code.

**Comment by ymjiang** [comment]
@yhpeng-git Merged. Thank you for your contributions!
